mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java

```diff
@@ -0,0 +1,420 @@
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select()
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+                .and(MESSAGE_UID.greaterOrEqual(range.getUidFrom().asLong()))
+                .and(MESSAGE_UID.lessOrEqual(range.getUidTo().asLong()))
+                .orderBy(DEFAULT_SORT_ORDER_BY)))
+            .map(recordToComposedMessageIdWithMetaData(mailboxId));
+    }
+
+    public Flux<ComposedMessageIdWithMetaData> findMessagesMetadata(PostgresMailboxId mailboxId, List<MessageUid> messageUids) {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select()
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+                .and(MESSAGE_UID.in(messageUids.stream().map(MessageUid::asLong).toArray(Long[]::new)))
+                .orderBy(DEFAULT_SORT_ORDER_BY)))
+            .map(recordToComposedMessageIdWithMetaData(mailboxId));
+    }
+
+
+    public Flux<ComposedMessageIdWithMetaData> findAllRecentMessageMetadata(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select()
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+                .and(IS_RECENT.eq(true))
+                .orderBy(DEFAULT_SORT_ORDER_BY)))
+            .map(recordToComposedMessageIdWithMetaData(mailboxId));
+    }
+
+    private Function<Record, ComposedMessageIdWithMetaData> recordToComposedMessageIdWithMetaData(PostgresMailboxId mailboxId) {
+        return record -> ComposedMessageIdWithMetaData
+            .builder()
+            .composedMessageId(new ComposedMessageId(mailboxId,
+                PostgresMessageId.Factory.of(record.get(MESSAGE_ID)),
+                MessageUid.of(record.get(MESSAGE_UID))))
+            .threadId(extractThreadId(record))
+            .flags(RECORD_TO_FLAGS_FUNCTION.apply(record))
+            .build();
+    }
+
+    public Mono<Void> updateFlag(PostgresMailboxId mailboxId, MessageUid uid, UpdatedFlags updatedFlags) {
+        return postgresExecutor.executeVoid(dslContext ->
+            Mono.from(buildUpdateFlagStatement(dslContext, updatedFlags, mailboxId, uid)));
+    }
+
+    public Mono<Flags> listDistinctUserFlags(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRow(dslContext -> Mono.from(dslContext.selectDistinct(USER_FLAGS)
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))))
+            .map(RECORD_TO_FLAGS_FUNCTION);
+    }
+
+    private UpdateConditionStep<Record> buildUpdateFlagStatement(DSLContext dslContext, UpdatedFlags updatedFlags,
+                                                                 PostgresMailboxId mailboxId, MessageUid uid) {
+        return Optional.of(dslContext.update(TABLE_NAME))
+            .map(updateStatement -> {
```

Comment:
write:

```suggestion
            .map(updateStatement -> applyFlag(updateStatement, IS_ANSWERED, Flag.ANSWERED)
```

And reduce the boiler plate